### PR TITLE
Update developer guide to include M1 Setup

### PR DESCRIPTION
**Pull request by ryanbogan**
### Description
Updates developer guide to include setup on Mac M1 machines
 
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/580
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -63,6 +63,11 @@ One easy way to install on mac or linux is to use pip:
 pip install cmake==3.23.3
 ```
 
+On Mac M1 machines, install cmake using:
+```bash
+brew install cmake
```
Comment (navneet1v) [inline DEVELOPER_GUIDE.md:68]: does pip doesn't work on this? also can we add the version of cmake while installing using brew?

File: DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -78,6 +83,44 @@ Additionally, the `gcc` toolchain needs to be installed on Mac. To install, run:
...
+```bash
+// Go to jni folder
+cd k-NN/jni
+
+// File changes required
+sed -i -e 's/\/usr\/local\/opt\/libomp\//\/opt\/homebrew\/opt\/llvm\//g' CMakeLists.txt
+sed -i -e 's/-march=native/-mcpu=apple-m1/g' external/nmslib/similarity_search/CMakeLists.txt
+sed -i -e 's/pragma message WARN/pragma message /g' external/nmslib/similarity_search/src/distcomp_scalar.cc
+sed -i -e 's/-mcpu=apple-a14/-mcpu=apple-m1/g' external/nmslib/python_bindings/setup.py
+sed -i -e 's/__aarch64__/__undefine_aarch64__/g' external/faiss/faiss/utils/distances_simd.cpp
```
Comment (navneet1v) [inline DEVELOPER_GUIDE.md:99]: have we thought around automating this change as part of gradle build/run?

File: DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -78,6 +83,44 @@ Additionally, the `gcc` toolchain needs to be installed on Mac. To install, run:
...
+```
+
+Next, obtain a minimum distribution tarball of the k-NN version you want to build:
+
+1. Fork the [OpenSearch Repo](https://github.com/opensearch-project/OpenSearch) into your github account.
+2. Clone the repository locally
+3. Run the following commands:
+```cd OpenSearch && ./gradlew -p distribution/archives/darwin-tar assemble```
+4. You should see a opensearch-min-<version>-SNAPSHOT-darwin-x64.tar.gz file present in distribution/archives/darwin-tar/build/distributions/
+5. Build k-NN by passing the OpenSearch distribution path in `./gradlew <integTest/run> -PcustomDistributionUrl="<Full path to .tar.gz file you noted above>"`
```
Comment (navneet1v) [inline DEVELOPER_GUIDE.md:122]: do we need customDistribution here? does OpenSearch is not providing the distribution for M1Mac?

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:68]: Pip didn't work for me when I was setting up my new laptop.

File: DEVELOPER_GUIDE.md (hunk 2 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:99]: If there is a way to detect specific architecture in the tasks, then we could change that.

Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:99]: I think that should be a separate issue though.

File: DEVELOPER_GUIDE.md (hunk 3 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:122]: Yeah it throws an error if you don't specify the exact path to the distribution:
```
Execution failed for task ':run'.
> Could not resolve all files for configuration ':opensearch_distro_extracted_testclusters--integTest-0-3.0.0-SNAPSHOT-'.
   > Could not resolve opensearch-distribution-snapshot:opensearch:3.0.0-SNAPSHOT.
     Required by:
         project :
      > Could not resolve opensearch-distribution-snapshot:opensearch:3.0.0-SNAPSHOT.
         > Could not get resource 'https://artifacts.opensearch.org/snapshots/core/opensearch/3.0.0-SNAPSHOT/opensearch-min-3.0.0-SNAPSHOT-darwin-arm64-latest.tar.gz'.
            > Could not HEAD 'https://artifacts.opensearch.org/snapshots/core/opensearch/3.0.0-SNAPSHOT/opensearch-min-3.0.0-SNAPSHOT-darwin-arm64-latest.tar.gz'. Received status code 403 from server: Forbidden

```

Comment (navneet1v) [reply DEVELOPER_GUIDE.md:122]: Thanks make sense. You can resolve this comment, but before that can you just create a github issue on the OpenSearch core repo to start creating the min distribution for M1Mac.

File: DEVELOPER_GUIDE.md (hunk 2 above)
Comment (navneet1v) [reply DEVELOPER_GUIDE.md:99]: Make sense. you can resolve this comment, and create a github issue for automating this.

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (navneet1v) [reply DEVELOPER_GUIDE.md:68]: make sense

**Review by navneet1v — DISMISSED** [review]

File: DEVELOPER_GUIDE.md (hunk 3 above)
Comment (heemin32) [reply DEVELOPER_GUIDE.md:122]: Could you try this? https://github.com/opensearch-project/k-NN/issues/580#issuecomment-1741272382

Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:122]: That link is just for the tarball, so we could provide the download.  However, the user would still have to specify the path to that download

**Review by heemin32 — DISMISSED** [review]

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:68]: After further investigation, it seems that homebrew doesn't have a built in way of installing an older version of libraries.  So the only option is to use `brew install cmake` to install the latest version, which is currently 3.27.7

File: DEVELOPER_GUIDE.md (hunk 3 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:128]: https://github.com/opensearch-project/OpenSearch/issues/10593

File: DEVELOPER_GUIDE.md (hunk 2 above)
Comment (ryanbogan) [reply DEVELOPER_GUIDE.md:105]: https://github.com/opensearch-project/k-NN/issues/1247

**Review by navneet1v — APPROVED** [review]

**Review by heemin32 — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -20,4 +20,5 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Infrastructure
 ### Documentation
 ### Maintenance
+- Update developer guide to include M1 Setup [#1222](https://github.com/opensearch-project/k-NN/pull/1222)
```
Comment (heemin32) [inline CHANGELOG.md:23]:
```suggestion
* Update developer guide to include M1 Setup [#1222](https://github.com/opensearch-project/k-NN/pull/1222)
```

**Review by martin-gaievski — APPROVED** [review]

**Review by naveentatikonda — APPROVED** [review]

**Pull request merged**